RODARS pipeline

gabrielwol opened the pull request:
## What this pull request accomplishes:

- Construction event data (RODARs)!
  - [`congestion_events.rodars_locations`](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-e0aefe2c7b7894915f8157dfa52a70909602f9f1842495dd630bbcc52444fbb8) is the main view we will be using!
- [`rodars_pull.py`](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-c5416de4fedc39cc36dc10184946f74ab78f9500d3d878640fc61abfd8153d9f) daily DAG which runs on Morbius.
  - Pulls RODARs data from ITSC in two parts:
    - [Issues](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-fdeea1442d17c80d90bac8b158669a5b4473061411ddf0c0640889416538c929): only the latest update (timestamputc) for each issueid. 
    - [Issue locations](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-b31f461bc75b89a15c4e9bc5dbc5b8c1894a4f44025d068e024aa8581ed5938c): corresponding location details for latest issues.  
  - Data is processed using functions in [`rodars_functions.py`](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-6e2d80f8ae4ae7abe1f10d795eedf9fe09d6195b6ea0c63f36193789655ff9ac), including: unnesting json data, converting binary coordinates to postgres readable format, data type conversion.
  - Finally, data is inserted into `congestion_events.rodars_issues` and `congestion_events.rodars_issue_locations`.
    - Notably older issue_locations are deleted using this [on insert trigger](https://github.com/CityofToronto/bdit_data-sources/pull/1103/files#diff-49aa6162dedc1c7f47675cfae6689e0569954666998e93d0be7b1a252df89983). I used this method because I thought the number of locations per issue could theoretically change, resulting in orphaned rows, unaffected by an `on conflict do update`.
- 5 new lookup tables taken from the documentation in `itsc_factors` schema: `direction`, `lanesaffectedpattern`, `locationblocklevel`, `roadclosuretype_new`, `roadclosuretype_old`. Used to convert codes to human readable format in `issue_locations` view.
  - Only the `itsc_factors.lanesaffectedpattern` lookup could use review: I devised the `lane_open`, `lane_closed` columns to give a numeric interpretation of how many lanes are open/closed. I used 0.5 for partial closures/slowdowns. 
  - Also checkout the function `itsc_factors.get_lanesaffected_sums` which translates these codes into numeric columns for ease of use in the main view. A `lanesaffectedpattern` could be something like `'LOLCLCWO'` (lane open, lane closed, lane closed, sidwalk open) (Try: `SELECT lane_open_auto, lane_closed_auto, lane_open_bike, lane_closed_bike, lane_open_ped, lane_closed_ped, lane_open_bus, lane_closed_bus FROM itsc_factors.get_lanesaffected_sums('LOLCLCWO')`)


## Issue(s) this solves:

- #1088

## What, in particular, needs to reviewed:

- [ ] Does this live in the right schema (`congestion_events`)? View/table names?
- [ ] Decision to only keep latest issue definition (every update is logged in ITS Central).
- [ ] Review the main view and the readme. 

## What needs to be done by a sysadmin after this PR is merged
- Refresh data_scripts on Morbius.

gabrielwol commented:
Ready to review! Will add some more usage examples to readme at some point.

radumas reviewed (changes requested):
Commenting on the documentation.

Pleased that the docs are extensive, you have thought out the limitations of the data, done some exploratory analysis and the use of the callouts for extra punchy information. 

I have some thoughts on re-ordering the README, but maybe we merge this and solicit feedback from users.

I have added some small editorial comments on language.

I think the big thing missing in terms of documentation is that parent folders should have a brief description of the contents of this folder so that people can know where to find it.

Also should this folder be more accurately called something like `lane_obstructions`? Or `road_permits`? Since the data are for events *and* construction (right?)

chmnata commented on events/construction/sql/create-table-itsc_factors_lanesaffectedpattern.sql:
```diff
@@ -0,0 +1,28 @@
+
+REVOKE ALL ON TABLE itsc_factors.lanesaffectedpattern FROM bdit_humans;
+
+GRANT SELECT ON TABLE itsc_factors.lanesaffectedpattern TO bdit_humans WITH GRANT OPTION;
```
without grant option

chmnata commented on events/construction/sql/create-function-get_lanesaffected_sum.sql:
```diff
@@ -0,0 +1,51 @@
+
+ALTER FUNCTION itsc_factors.get_lanesaffected_sums(text) OWNER TO congestion_admins;
+
+GRANT EXECUTE ON FUNCTION itsc_factors.get_lanesaffected_sums(text) TO public;
```
bdit human instead of public?

chmnata commented on events/construction/sql/create-view-rodars_locations.sql:
```diff
@@ -0,0 +1,120 @@
+    CASE iil.centreline_id WHEN 0 THEN NULL ELSE iil.centreline_id END AS centreline_id,
+    COALESCE(
+        centreline_latest.geom,
+        --find geoms for centreline that do not appear in centreline_latest
```
ohhhh nice

chmnata commented on events/construction/sql/insert-rodars_issue_locations.sql:
```diff
@@ -0,0 +1,69 @@
+    roadclosuretype,
+    st_geomfromtext(geom_text, 4326) AS geom
+FROM locations;
+--ON CONFLICT (divisionid, issueid, locationindex, direction)
```
delete if not needed anymore?

gabrielwol commented:
@radumas readme changes ready for re-review:
- renamed folder to road_permits, moved existing contents of events into events/special_events and added events level readme. 
- updated reference in top level data-sources readme + removed old `road_closure` folder

radumas reviewed (approved):
Looks great! Thank you!

gabrielwol commented:
> Is there a bug in the linting action that it isn't annotating?

I also noticed previously on this PR there were some lines being erroneously annotated with line length sqlfluff error. Perhaps time to look into upgrading sqlfluff action.
